k8s/cmd/report.go

```diff
@@ -0,0 +1,110 @@
+package cmd
+
+import (
+	"fmt"
+
+	"github.com/iter8-tools/iter8/base/log"
+	basecli "github.com/iter8-tools/iter8/cmd"
+
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
+
+type ReportOptions struct {
+	Streams              genericclioptions.IOStreams
+	ConfigFlags          *genericclioptions.ConfigFlags
+	ResourceBuilderFlags *genericclioptions.ResourceBuilderFlags
+	namespace            string
+	client               *kubernetes.Clientset
+
+	experimentId string
+}
+
+func newReportOptions(streams genericclioptions.IOStreams) *ReportOptions {
+	rbFlags := &genericclioptions.ResourceBuilderFlags{}
+	rbFlags.WithAllNamespaces(false)
+
+	return &ReportOptions{
+		Streams:              streams,
+		ConfigFlags:          genericclioptions.NewConfigFlags(true),
+		ResourceBuilderFlags: rbFlags,
+	}
+}
+
+// complete sets all information needed for processing the command
+func (o *ReportOptions) complete(factory cmdutil.Factory, cmd *cobra.Command, args []string) (err error) {
+
+	o.namespace, _, err = factory.ToRawKubeConfigLoader().Namespace()
+	if err != nil {
+		return err
+	}
+
+	o.client, err = GetClient(o.ConfigFlags)
+	if err != nil {
+		return err
+	}
+
+	if len(o.experimentId) == 0 {
+		s, err := GetExperimentSecret(o.client, o.namespace, o.experimentId)
+		if err != nil {
+			return err
+		}
+		o.experimentId = s.Labels[IdLabel]
+	}
+
+	return err
+}
+
+// validate ensures that all required arguments and flag values are provided
+func (o *ReportOptions) validate(cmd *cobra.Command, args []string) (err error) {
+	return nil
+}
+
+// run runs the command
+func (o *ReportOptions) run(cmd *cobra.Command, args []string) (err error) {
+	expIO := &KubernetesExpIO{
+		Client:    o.client,
+		Namespace: o.namespace,
+		Name:      SpecSecretPrefix + o.experimentId,
+	}
+
+	log.Logger.Trace("build started")
+	exp, err := basecli.Build(true, expIO)
+	log.Logger.Trace("build finished")
+	if err != nil {
+		return err
+	}
+
+	return exp.Report(basecli.ReportOptions.OutputFormat)
+}
+
+func NewReportCmd(factory cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+	o := newReportOptions(streams)
+
+	cmd := basecli.NewReportCmd()
+	var example = `
+# Generate text report for the most recent experiment running in current Kubernetes context
+iter8 report --remote`
+	cmd.Example = fmt.Sprintf("%s%s\n", cmd.Example, example)
+	cmd.RunE = func(c *cobra.Command, args []string) error {
+		if err := o.complete(factory, c, args); err != nil {
```

Comment:
PreRunE is a better place for `o.complete` and `o.validate`, as opposed to `RunE`, which can contain only the run logic. 

Incidentally, what does `o.complete` do?

Reply:
`complete()` is computing value of variables used by `run()` that either need default values or that can be derived from the input. Most of these are the same; perhaps a common method can be used.

Reply:
I saw the complete() code which either is no-op in one case, or replicated in other cases.

So, yes, removing the code replication and making a single function makes sense.

I now understand what complete does. You can keep this in RunE or PreRunE depending upon your preference.